### PR TITLE
Add gcc6 plugin

### PR DESCRIPTION
**Pull request by tehnick**

### CODE REVIEW

**Comment by tehnick** [comment]
I have tested qt5, openscenegraph and freeglut with gcc 6.1.0 in my customized mxe fork:
all is built and works fine.

**Comment by starius** [comment]
Why have you replaced HTTP URL with FTP URL?
FTP servers, especially FTP servers of GNU are not ideal.
I have fixed all FTP URLs in https://github.com/mxe/mxe/pull/435

**Comment by starius** [comment]
Comment "libmysqlclient patch changes don't adversely affect 5.2 series" seems to be related to gcc 5, but it was copied to gcc 6 plugin.

**Comment by tehnick** [comment]
> Why have you replaced HTTP URL with FTP URL?

http://isl.gforge.inria.fr/ was unavailable yesterday during all my working day, so I had to fix mxe build somehow. But now this site is alive. So my impression about its death was premature...

**Comment by starius** [comment]
I'm building all the packages using the following command:
```
$ MXE_PLUGIN_DIRS=plugins/gcc6 lua tools/build-pkg.lua
```

The results are expected in 1-2 days.

**Comment by starius** [comment]
> http://isl.gforge.inria.fr/ was unavailable yesterday during all my working day

It was chosen almost two years ago. Now it is not even listed in https://gcc.gnu.org/mirrors.html Maybe it is time to choose other mirrors?

You can specify two servers:
```
$ grep URL plugins/native/automake.mk
$(PKG)_URL      := http://ftp.gnu.org/pub/gnu/automake/$($(PKG)_FILE)
$(PKG)_URL_2    := ftp://ftp.gnu.org/pub/gnu/automake/$($(PKG)_FILE)
```

**Comment by tehnick** [comment]
All remarks are fixed.

**Comment by tehnick** [comment]
> It was chosen almost two years ago. Now it is not even listed in https://gcc.gnu.org/mirrors.html Maybe it is time to choose other mirrors?

As far as I know isl is not a gnu project, so it is not hosted on usual gnu mirrors. But I may be mistaken.

**Comment by tehnick** [comment]
> FTP servers, especially FTP servers of GNU are not ideal.
> I have fixed all FTP URLs in #435

It is very strange, because they use strict criteria for hosting of their projects:
https://www.gnu.org/software/repo-criteria.en.html

**Comment by starius** [comment]
> It is very strange, because they use strict criteria for hosting of their projects:

Nevertheless their FTP servers ban (or used to ban) some IP addresses. Moreover, some machines are located behind a firewall which breaks FTP protocol.

**Comment by starius** [comment]
The build is still in progress.
Some packages failed:

  * boost
  * fdk-aac
  * flann
  * glib
  * itk
  * jsoncpp
  * qt
  * smpeg2
  * ucl

Logs: https://gist.github.com/starius/81e25169242155aa3ef6be1a733b9812
Global log (part): https://gist.github.com/c01ef084eeb85781bd1eb7f6b1e12192

**Comment by tehnick** [comment]
> Some packages failed:

It was expected due to:
"The C++ frontend now defaults to C++14 standard instead of C++98 it has been defaulting to previously, for compiling older C++ code that might require either explicitly compiling with selected older C++ standards, or might require some code adjustment..."

More details:
https://gcc.gnu.org/ml/gcc-announce/2016/msg00000.html
https://gcc.gnu.org/gcc-6/changes.html

But should it prevent merging this pull request?

As a workaround we may add -std=c++11 or -std=gnu++98 into CXXFLAGS for building problematic packages. And this will not affect the builds with older versions of gcc.

**Comment by tonytheodore** [comment]
> But should it prevent merging this pull request?

I don't think so, the sooner it's available, the sooner people can start contributing fixes for broken packages and detecting other more subtle runtime issues.

> As a workaround we may add -std=c++11 or -std=gnu++98 into CXXFLAGS for building problematic packages. And this will not affect the builds with older versions of gcc.

Sounds like a good idea - can you add a `README.md` for this plugin that has links to the failing logs and notes above? It can be useful as a "status" and tips page for people wanting to try the latest version.

**Comment by tehnick** [comment]
> Sounds like a good idea - can you add a README.md for this plugin that has links to the failing logs and notes above?

Yes, sure.

**Comment by tehnick** [comment]
Done.

**Comment by tehnick** [comment]
+ first fixes in broken packages

**Comment by tehnick** [comment]
@starius Any updates in your list of broken packages?

**Comment by starius** [comment]
@tehnick, I am building commit 2d2e1f8b89c9551948935647c09e2dda509f5bcc

It is not finished, but some packages are already broken:

```
[build-pkg]     Item is broken: x86_64-w64-mingw32.shared~jsoncpp
[build-pkg]     Item is broken: x86_64-w64-mingw32.shared~fdk-aac
[build-pkg]     Item is broken: x86_64-w64-mingw32.shared~flann
[build-pkg]     Item is broken: x86_64-w64-mingw32.shared~json_spirit
[build-pkg]     Item is broken: x86_64-w64-mingw32.shared~libxml++
[build-pkg]     Item is broken: x86_64-w64-mingw32.shared~itk
[build-pkg]     Item is broken: x86_64-w64-mingw32.shared~cgal
[build-pkg]     Item is broken: x86_64-w64-mingw32.shared~vtk6
[build-pkg]     Item is broken: x86_64-w64-mingw32.shared~ossim
[build-pkg]     Item is broken: x86_64-w64-mingw32.shared~qt3d
[build-pkg]     Item is broken: x86_64-w64-mingw32.shared~opencv
```

Logs: https://gist.github.com/f0a78ec2f1fbe92d92f250e4ff22bc51

**Comment by tehnick** [comment]
@starius 
Hmm, I have not tested shared target. Ok, I'll look at broken packages which I have fixed for static target.

**Comment by tehnick** [comment]
@starius @tonytheodore
Is it possible that CXXFLAGS variable is cached anywhere in MXE?

For example, in file https://github.com/mxe/mxe/pull/1317/commits/2fe841cd9380b76fbf3d3f4c02bda4a1681ffebf#diff-24a148f8682f585541ad471cfbecc142R14 I explicitly set
`CXXFLAGS := -Wno-error=conversion -Wno-error=shift-negative-value`
and
`-DCMAKE_CXX_FLAGS="$(CXXFLAGS)"`.
And it worked earlier...

But after building other packages when I rebuild `jsoncpp` it FTBFS and I see in logs very strange thing: `... -DCMAKE_CXX_FLAGS="-std=gnu++11" ...` which is should not be there.

Have you got any ideas?

My knowledge of Makefile syntax is quite limited and probably I have missed something obvious...

**Comment by tonytheodore** [comment]
Variable scope can be a little tricky to follow as the contents of `$(PKG)_BUILD` are `eval`ed after Make's normal two-phase parsing.

Most Make variables are global and `CXXFLAGS` is being defined multiple times with it's final value being taken from the last file parsed. If you are working your way through packages alphabetically, it will appear to work as the last value will be the one you just set, but subsequent runs will use the final value (from `ucl` in this case).

The two simplest ways to set this correctly are:
 - don't use a Make variable and set CXXFLAGS env variable directly in build rule (as you've done with `qt.mk`)
 - give the Make variable a unique name 
```
# workaround for builds with GCC >= 6.x
$(PKG)_CXXFLAGS := -std=gnu++11
...
define $(PKG)_BUILD
    cd '$(1)' && \
        CXXFLAGS="$($(PKG)_CXXFLAGS)" ...
```

Note that in the second case, `$(PKG)` acts as both a normal variable with global scope for the initial parsing of `$(PKG)_CXXFLAGS`, then as a target-specific variable (defined at runtime in the main `Makefile`).

**Comment by starius** [comment]
The build is still in progress. Some static packages are broken:

```
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~flann
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~ucl
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~jsoncpp
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~freeimage
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~libical
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~wxwidgets
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~fdk-aac
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~itk
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~json_spirit
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~dcmtk
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~libxml++
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~cgal
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~vtk6
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~vtk
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~ossim
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~librsvg
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~qt3d
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~sdl_sound
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~opencv
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~gtkmm3
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~gtkmm2
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~gtkimageview
[build-pkg]     Item is broken: x86_64-w64-mingw32.static~log4cxx
[build-pkg]     Item is broken: i686-w64-mingw32.static~jsoncpp
[build-pkg]     Item is broken: i686-w64-mingw32.static~fdk-aac
[build-pkg]     Item is broken: i686-w64-mingw32.static~ucl
[build-pkg]     Item is broken: i686-w64-mingw32.static~freeimage
[build-pkg]     Item is broken: i686-w64-mingw32.static~json_spirit
```

**Comment by tehnick** [comment]
@starius
Thanks. I have updated README.md

@tonytheodore
Thank you for explanation.

I have updated commit for jsoncpp and have tested builds of glib, jsoncpp, boost, qt, smpeg2 and smpeg packages for x86_64-w64-mingw32.static and x86_64-w64-mingw32.shared. They are fine now.

I have dropped commit for ucl package because it needs more attention for proper fix.

Also I have rebased this pull request against current master branch, so it should be ready for merging.

I will try to fix a few more packages later (in another pull request) when I will find some free time.

File: plugins/gcc6/README.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+...
+$(PKG)_CXXFLAGS := -std=gnu++11
+...
+    cd '$(1)' && \
+    CXXFLAGS="$($(PKG)_CXXFLAGS)" \
+    ./configure \
+...
+```
+
+Just after adding this plugin (gcc6) some packages were FTBFS:
```
Comment (starius) [inline plugins/gcc6/README.md]: IMHO, s/were FTBFS/failed to build from source/
It was hard to find the meaning :(

**Comment by starius** [comment]
> I have updated commit for jsoncpp and have tested builds of glib, jsoncpp, boost, qt, smpeg2 and smpeg packages for x86_64-w64-mingw32.static and x86_64-w64-mingw32.shared. They are fine now.

The list of broken packages in README.md still contains them. I think, only packages broken after this pull request should be listed there.

The pull request looks good to me. I am rebuilding all packages from scratch for 57cb6bb571d650feb5549e88cbbdf2c4dc18b878 with and without the plugin.

File: plugins/gcc6/README.md (hunk 1 above)
Comment (tehnick) [inline plugins/gcc6/README.md]: No, this abbreviation is explained in the begging of README.

**Comment by tehnick** [comment]
> The list of broken packages in README.md still contains them. I think, only packages broken after this pull request should be listed there.

I disagree. Fixed packages also should be present in the list as examples how packages could be fixed.

> The pull request looks good to me. I am rebuilding all packages from scratch for 57cb6bb with and without the plugin.

Hmm, this is good idea. I'll check fixed packages without plugin as well.

**Comment by starius** [comment]
> I disagree. Fixed packages also should be present in the list as examples how packages could be fixed.

I see. Maybe the following table is better:

| package            | target               | fixed in commit        |
| ------------------ | -------------------- | ---------------------- |
| boost              | all                  | 7ca2bceedd41c0919a30f9 |
| ucl                | all                  | -                      |
| ...                | ...                  | ...                    |

Source:
```markdown
| package            | target               | fixed in commit        |
| ------------------ | -------------------- | ---------------------- |
| boost              | all                  | 7ca2bceedd41c0919a30f9 |
| ucl                | all                  | -                      |
| ...                | ...                  | ...                    |
```

**Comment by tehnick** [comment]
> Maybe the following table is better:

Yes, it looks nice.

**Comment by tehnick** [comment]
Done.

**Comment by starius** [comment]
GitHub doesn't convert commits to links at
https://github.com/tehnick/mxe/blob/gcc6-plugin/plugins/gcc6/README.md
They are too short or it just doesn't support highlighting commits in README pages.

**Comment by tehnick** [comment]
Look now.

**Comment by starius** [comment]
README.md looks good to me.

**Comment by tonytheodore** [comment]
Perfect, thanks!

**Pull request merged**